### DISCUSSION TITLE
imageanalysis

### DISCUSSION THREAD

**Discussion by petermr** [petermr/petermr#8] [Announcements]
This thread is for all discussions/ recording about the  https://github.com/petermr/pyamiimage repository. 

Currently Anuv (https://github.com/anuvc)  and I are developing `pyamiimage` to completely interpret a scientific diagram. In simple terms:
* Anuv is developing the text aspect (using Tesseract to extract strings and their coordinates and then creating textboxes with words,, phrases and paragraphs
* I (@petermr ) am extracting non-text primitives (lines, curves, maybe areas) and building higher-level primitives (arrows, straight lines, circles, boxes) 
* we then merge these approaches to create higher-level concepts (legends, axes, reactions)

**Comment by petermr** [comment]
# AmiImage

Having run into problems I plan to revamp the way images are processed and stored. We have to support:
* pipeline (often in this order)
- RGBA
- RGB
- grayscale
- binary (black/white)
- skeleton
* optional preprocessing:
- sharpening
- colour octrees

in `ami3` I used the filestore to support this. I think now I will use a mixture of a `dict` and filestore. The `dict` will allow lazy processing and caching.

**Comment by petermr** [comment]
# skeleton and nx_graph code.

The current integrated approach is to use `AmiGraph.create_ami_graph_from_arbitrary_image_file()`
```
    def test_skeleton_for_plot(self):
        """create  skeleton for an xy-plot
        """
        infile = Resources.YW5003_5
        image0 = io.imread(str(Path(infile)))
        print(f"{infile} shape {image0.shape}")
        ami_graph = AmiGraph.create_ami_graph_from_arbitrary_image_file(infile)
        assert ami_graph is not None
        islands = ami_graph.get_or_create_ami_islands()
        assert len(islands) == 227, f"islands from IUCr plot , found {len(islands)}"
        for island in islands:
            bbox = island.get_or_create_bbox()
            if bbox.max_dimension() > 50:
                print(f"island {bbox} {len(island.node_ids)} {island.node_ids}")

```
gives the analysis:
```
/Users/pm286/workspace/pyamiimage/test/resources/iucr_yw5003_fig5.png 
shape (828, 1063, 4)
island [[341, 664], [49, 763]] 96 {12, 13, 14, 15, 16, 17, 18, 19, 20, 525, 23, 26, 29, 30, 31, 32, 33, 34, 35, 38, 41, 55, 567, 568, 58, 570, 571, 572, 573, 574, 575, 576, 578, 579, 580, 581, 582, 583, 577, 73, 586, 585, 76, 588, 589, 584, 587, 83, 84, 597, 598, 85, 87, 88, 89, 91, 604, 601, 602, 95, 600, 603, 99, 118, 130, 151, 152, 192, 215, 287, 296, 339, 343, 344, 348, 349, 352, 355, 357, 358, 359, 360, 361, 362, 365, 383, 385, 386, 390, 407, 449, 450, 451, 599, 469, 511}
island [[65, 295], [61, 770]] 36 {258, 132, 389, 136, 24, 288, 546, 36, 40, 424, 46, 177, 569, 63, 448, 193, 323, 590, 591, 592, 593, 594, 595, 596, 86, 605, 606, 607, 608, 609, 354, 101, 103, 487, 114, 510}
```

**Reply by petermr** [reply]
# island size analysis 

A variety of size filters for islands. "very small", "very large", "long and thing",  "short and fat", etc. Not fully tested:
```
in test_ami_graph.py:

    def test_island_sizes(self):
        """uses mindim, maxdim, to filter in/out islands. etc.
        """
        ami_graph = AmiGraph.create_ami_graph_from_arbitrary_image_file(Resources.YW5003_5);
        #all islands
        islands = ami_graph.get_or_create_ami_islands()
        assert len(islands) == 227, f"expected total islands {len(islands)}"
        # two largest
        islands = ami_graph.get_or_create_ami_islands(mindim=50)
        assert len(islands) == 2, f"expected large islands {len(islands)}"
        # very small ones (periods, dots, etc.)
        islands = ami_graph.get_or_create_ami_islands(maxdim=2)
        assert len(islands) == 16, f"very small islands (periods, dots, decimals) {len(islands)}"
        # all thin , including dots
        islands = ami_graph.get_or_create_ami_islands(maxmindim=3)
        assert len(islands) == 49, f"very thin islands (sans-serif l, i, etc) {len(islands)}"
        # long thin , excluding dots
        islands = ami_graph.get_or_create_ami_islands(minmaxdim=19)
        assert len(islands) == 22, f"thin long islands (sans-serif l, i, etc) {len(islands)}"
        # long thin , excluding dots
        islands = ami_graph.get_or_create_ami_islands(minmaxdim=24)
        assert len(islands) == 9, f"very thin long islands (sans-serif l, i, etc) {len(islands)}"
        # islands with no thin dimension
        islands = ami_graph.get_or_create_ami_islands(mindim=15)
        assert len(islands) == 21, f"chunky islands (H M O etc and the 2 boxes) {len(islands)}"

```

**Comment by petermr** [comment]
# extraction of plot components

NOTE: Do NOT do store individual numpy values for coordinates as they are not Python ints. Use `AmiUtil.get_xy_from_sknw_centroid`.

This example uses 
![iucr_yw5003_fig5](https://user-images.githubusercontent.com/733515/148706273-f6375932-a1bf-4780-8449-e6ac6aed90cf.png) (left-hand plot)


```
    def test_graph_plots(self):
        """uses mindim, maxdim, to filter in/out islands. etc.
        """
        ami_graph = AmiGraph.create_ami_graph_from_arbitrary_image_file(Resources.YW5003_5);
        # second largest island is a boxed plot
        #all islands
        islands = ami_graph.get_or_create_ami_islands(mindim=50, maxmindim=300)
        assert len(islands) == 1, f"expected single islands {len(islands)}"
        small_plot = islands[0]
        node_ids = small_plot.node_ids
        assert len(node_ids) == 36, f"nodes in small graph {len(node_ids)}"
        assert node_ids == {258, 132, 389, 136, 24, 288, 546, 36, 40, 424, 46, 177, 569, 63, 448, 193,
                         323, 590, 591, 592, 593, 594, 595, 596, 86, 605, 606, 607, 608, 609, 354,
                         101, 103, 487, 114, 510}
        assert small_plot.edges == [(258, 323), (132, 177), (389, 448), (136, 177), (136, 193), (24, 36),
                                    (24, 36), (24, 40), (288, 323), (288, 354), (546, 569), (546, 595),
                                    (36, 46), (40, 46), (40, 63), (424, 448), (424, 487), (46, 86),
                                    (177, 258), (569, 590), (63, 86), (63, 101), (448, 510), (193, 258),
                                    (193, 288), (323, 389), (590, 591), (590, 608), (591, 592), (591, 605),
                                    (592, 593), (592, 606), (593, 594), (593, 596), (594, 595), (594, 607),
                                    (595, 609), (86, 103), (354, 389), (354, 424), (101, 103), (101, 114),
                                    (103, 132), (487, 510), (487, 546), (114, 132), (114, 136), (510, 569)]
        for node_id in node_ids:
            ami_node = AmiNode(node_id=node_id, ami_graph=ami_graph)
            node_xy = ami_node.centroid_xy

            for neighbour_id in ami_node.get_neighbors():
                neighbour_xy = AmiNode(node_id=neighbour_id, ami_graph=ami_graph).centroid_xy
                if abs(neighbour_xy[0] - node_xy[0]) <= 2:
                    # vert line
                    pass
                elif abs(neighbour_xy[1] - node_xy[1]) <= 2:
                    # horiz line
                    pass
                else:
                    # only count once
                    if neighbour_xy[0] < node_xy[0]:
                        print(f"y plot  {node_id}..{neighbour_id} => {node_xy}...{neighbour_xy}")
```
Console output:
```
Launching pytest with arguments pyamiimage/test/test_ami_graph.py::TestAmiGraph::test_graph_plots --no-header --no-summary -q in /Users/pm286/workspace

============================= test session starts ==============================
collecting ... collected 1 item

pyamiimage/test/test_ami_graph.py::TestAmiGraph::test_graph_plots 

======================= 1 passed, 140 warnings in 10.89s =======================

Process finished with exit code 0
PASSED [100%]y plot  136..177 => [295, 372]...[66, 408]
y plot  24..36 => [295, 96]...[66, 131]
y plot  288..323 => [295, 483]...[66, 520]
y plot  546..569 => [295, 715]...[66, 752]
y plot  546..595 => [295, 715]...[274, 759]
y plot  40..46 => [295, 151]...[66, 185]
y plot  424..448 => [295, 598]...[66, 632]
y plot  63..86 => [295, 206]...[66, 241]
y plot  193..258 => [295, 427]...[66, 462]
y plot  354..389 => [295, 542]...[66, 577]
y plot  101..103 => [295, 267]...[66, 300]
y plot  487..510 => [295, 655]...[66, 692]
y plot  114..132 => [295, 324]...[66, 355]
```
These are the 12 R-L plot traces (and the bottom RH corner 546-95.

Next stage is to extract the coords into X-Y files.

**Comment by anuvc** [comment]
## Tesseract OCR -> AmiOCR
I am redoing tesseract into a new class with better data structure management called AmiOCR.

### AmiOCR
AmiOCR objects will contain information about the text boxes in an image, with TextBox objects.

### TextBox
TextBox class inherits from BBox, so it is a bounding box with a text in it.

**Comment by petermr** [comment]
# Geometry class

I need a class to manage geometry, e.g. intersection of lines. I have been pointed at Shapely so will start with that. 
Also Sympy. https://docs.sympy.org/latest/modules/geometry/index.html 
So the extracted straight lines will be converted to or wrapped in Shapely.
or Sympy!
Sympy does 3D which is very useful for other disciplines.

**Comment by petermr** [comment]
# Constructing boxes and axial info

The extracted nodes and segments must be reassembled as axes with tick marks . This needs to be general (i.e. nor rewritten for each  case. My current thoughts:
* extract all nodes and edges within a border of the box 
* identify the nodes (1- and 3-, possibly 4- at the corners)
* turn T-joints (with short perpendicular) into **separate** lines . Joint the longer straight lines into a single line (iteratively). This is only needed on the bottom edge of this one.
* separate T-joints with 2 straight lines iand 1 curved into a single line and a curved line
* create corners using Douglas-Peucker (delta <=  2 max). 
* add 2- nodes at corners
* identify 4 corners and check metrics for rectangle

We should now have:
* one box
* one set of parallel short ticks
* 12 curved lines

**Comment by petermr** [comment]
# nodes and edges in plots

The terminology/syntax can be complex as we inherit the NetworkX`nx_graph`  from `sknw`.
We use the SKNW Multigraph option creates a NetworkX Multigraph (where there can be multiple edges between nodes). This means that every node-node connection needs a branch number (branch_id).
```
nx_graph[start_id][end_id][branch_id] # this is an edge between the nodes
```
Note that in a non-directed graph every edge is referenced in both directions:
```
nx_graph[end_id][start_id][branch_id] # this references the reverse edge
```

to address nodes we use:
```
nx_graph.nodes
```
rather than 
```
nx_graph[2]
```

 
        Note:
        `nx_graph[i][j]` seens to be the same as `nx_graph[j][i]` in undirected graphs but may not be the same object

Using:
Resources.BIOSYNTH1_ARROWS

![Screenshot 2022-01-14 at 10 49 55](https://user-images.githubusercontent.com/733515/149503795-8c60486b-527b-4f84-9bb4-dfd7164ff573.png)


We can skeletonize and create an `sknw` `nx_graph` with 
```
        nx_graph = AmiGraph.create_nx_graph_from_arbitrary_image_file(Resources.BIOSYNTH1_ARROWS)
```

        {0, 1, 2, 3, 4, 5, 6, 7},  # double arrow
            0         6
          /            \\
    3----2-------------4----5
          \\            /
           7          1
         [(0, 2), (1, 4), (2, 4), (2, 3), (2, 7), (4, 5), (4, 6),
        """
        print("\n2 0", nx_graph[2][0][0]["pts"][:2:-2])

```
        points0_2 = nx_graph[2][0][0]["pts"]
        assert str(points0_2) == "2 0 [[ 91 857]\
 [ 88 855]\
 [ 87 853]\
 [ 86 851]\
 [ 85 849]\
 [ 84 847]]\
        ]", f"points [2][0][0]['pts'] should be {str(points0_2)} "
```
These coordinates are `[y, x]` and we (may) need to change the sign

**Comment by anuvc** [comment]
## Development Update

### `AmiOCR` objects
`AmiOCR` objects store all the textual data in an image. 
For simple pathway images, this object stores the text (string) location (`TextBox` object) of words, phrases and 'word groups'.

`AmiOCR` object accepts a `Path` object (from pathlib) or a string of the path of the image you want to OCR.

#### Usage
```
obj = AmiOCR(image_path)
words = obj.get_words()
```
#### Example 
```  
biosynth2_path = Path("resources/biosynth2.png")  
biosynth2_text = AmiOCR(biosynth2.png)  

words = biosynth2_text.get_words()
```

### TODOs
- Currently `get_words()` work, however, `get_phrases()` and `get_groups()` are yet to be completed. 
- A filter to remove errors made during OCR, such as misidentification of graphical primitives as special characters.
- AmiOCR handler for numpy arrays of images.

**Comment by anuvc** [comment]
### Added plotting bounding boxes to AmiOCR

AmiOCR stores text words in TextBox objects which inherit BBox object. I have added a classmethod in BBox which plots a rectangle in the image by setting all the pixels on the rectangle to be 0 (black).

There are a few bugs that I'll iron out like: the words at the edges seem to have a bounding box which goes beyond the image, as a result python is throwing an IndexError. I have put a weak solution for now, ignoring such illegal bounding boxes


![biosynth2_snippet](https://user-images.githubusercontent.com/45652303/150409760-5a5c2f64-1826-469a-a281-0750753ef6d3.png)

**Comment by petermr** [comment]
# Segmenting lines

(This was a bits of a struggle and I thought it wasn't working but it seems good now!)

So far we can create nodes and edges which we wrap as `AmiNode` and `AmiEdge`. These contain the NetworkX/sknw nodes and edges but add convenience methods to avoid errors and also add more functionality. An AmiEdge is indexed in AmiGraph, not nx.graph. All graphs are **multigraphs**.

Each edge contains a triple `(start_id, end_id, branch_id)` where `branch_id` is a zero-based index of the edge between the nodes. I hope all duples have been replaced. The user should be able to iterate over all edges between a pair of nodes.

We extract the points from each AmiEdge (wrapping the `[s][e][b]["PTS"]` in convenience functions). The `sknw` output is `(y,x)` and we transform to `x,y` in a `numpy` array of `(N, 2)` .

This is passed to `skimage` segmentation. The segments are returned a `AmiLine`s, a new class which currently only contains start and end coordinates.
```
    def test_create_line_segments(self):
        """tests straightness between nodes (horiz and vert)
        """
        ami_graph = AmiGraph.create_ami_graph_from_arbitrary_image_file(Resources.YW5003_5)
        small_plot_island = ami_graph.get_or_create_ami_islands(mindim=50, maxmindim=300)[0]
        ami_edges = small_plot_island.get_or_create_ami_edges()
        tolerance = 1
        for ami_edge in ami_edges:
            ami_lines = ami_edge.get_ami_lines(tolerance)
            print(f"ami_lines (coords) ", ami_lines, "\n=========================\n")
```
This creates an AmiIsland for the LH plot ("(a)", see above) and generates 48 sets of AmiEdges . The order is arbitrary. Here are some commented examples:

The plot runs from y=61-> 764, x=66->295

```
pyamiimage/test/test_ami_graph.py::TestAmiGraph::test_create_line_segments 

======================= 1 passed, 140 warnings in 12.34s =======================

Process finished with exit code 0
PASSED [100%]
vector (0, 58)
ami_lines (coords)  [[[66, 462], [66, 520]]] 
=========================

```
This is the vertical chunk of the LH spine near the "N" of "Normalised". It's a single line of length 58,
```
vector (0, 55)
ami_lines (coords)  [[[66, 577], [66, 632]]] 
=========================

vector (-63, 1)
vector (-26, -3)
vector (-34, -1)
vector (-12, -3)
vector (-11, -5)
vector (-10, 0)
vector (-8, 7)
vector (-11, 15)
vector (-14, 12)
vector (-24, 10)
vector (-16, 3)
ami_lines (coords)  [[[295, 372], [232, 373]], [[232, 373], [206, 370]], [[206, 370], [172, 369]], [[172, 369], [160, 366]], [[160, 366], [149, 361]], [[149, 361], [139, 361]], [[139, 361], [131, 368]], [[131, 368], [120, 383]], [[120, 383], [106, 395]], [[106, 395], [82, 405]], [[82, 405], [66, 408]]] 
```
This is a roughly horizontal curve ending at the "s" of "Normalised". The algorithm has broken it into 11 segments none of which are more that 1 pixel away from the curve.
```
vector (-1, -35)
vector (-228, 0)
vector (0, 70)
ami_lines (coords)  [[[295, 96], [294, 61]], [[294, 61], [66, 61]], [[66, 61], [66, 131]]] 
```
This is the "top" of the diagram consisting of 3 lines meeting at right angles.
```
vector (41, 1)
ami_lines (coords)  [[[66, 758], [107, 759]]] 
=========================
```
part of the lower x-axis spine

**Comment by petermr** [comment]
# Segmentation of Edges

(NOTE: I'll use `AmiEdge` for a  set of points connecting 2 `AmiNode`s and `AmiLine` for any straight segment of that edge. "line" may be confusing).

Here's the current code (in test_ami_graph.test_create_line_segments)

```
   def test_create_line_segments(self):
        """segments the edge into straight-lines (AmiLine) and finds axially aligned corners
        """
        ami_graph = AmiGraph.create_ami_graph_from_arbitrary_image_file(Resources.YW5003_5)
        small_plot_island = ami_graph.get_or_create_ami_islands(mindim=50, maxmindim=300)[0]
        ami_edges = small_plot_island.get_or_create_ami_edges()
        assert len(ami_edges) == 48, f"found {len(ami_edges)}"
        tolerance = 2  # crude horiz-vert algorith does not yet use slopes
        for ami_edge in ami_edges:
            print(f"====edge==={ami_edge.get_id()}")
            lines_corners = ami_edge.find_single_lines_and_corners(tolerance)
            if not lines_corners:
                pass
            elif len(lines_corners) == 1:  # single line (not necessarily horizontal or vertical)
                print(f"line: {lines_corners[0]}")
            elif len(lines_corners) > 0:  # 20 or more lines all aligned with axes
                print(f"lines with axial corners: {lines_corners}")
```

gives (note the edges use (start_end_branch) ids, the lines use [x1,y1],[x2,y2]
```
pyamiimage/test/test_ami_graph.py::TestAmiGraph::test_create_line_segments 

======================= 1 passed, 140 warnings in 8.88s ========================

Process finished with exit code 0
====edge===258_323_0
line: [[66, 462], [66, 520]].   # single vertical line
...
====edge===136_177_0
# 2 or more segments with some non-axial
...
====edge===24_36_0
lines with axial corners: [[[295, 96], [294, 61]], [[294, 61], [66, 61]], [[66, 61], [66, 131]]]
```
This is 3 axial lines (vert-hor-vert) + 2 corners at the top of the diagram
```
====edge===546_595_0
lines with axial corners: [[[295, 715], [295, 757]], [[295, 757], [274, 759]]]
```
This is the bottom right corner vert-hor
```
====edge===591_592_0
line: [[107, 759], [149, 759]]
```
This is a short horizontal line (inter-tick)

**Reply by petermr** [reply]
## Axial lines

Now we extract the horizontal and vertical lines. These are important in most diagrams - they form axes, rectangles, underlines, etc.  Many of the `AmiEdge`s are single lines aligned with the axes (with associated `AmiNode`s at their ends). NOTE: the axial polylines are not extracted yet - this is the next phase.
```
    def test_filter_line_segments(self):
        """filters segments the edge into straight-lines (AmiLine) and finds axially aligned corners
        """
        ami_graph = AmiGraph.create_ami_graph_from_arbitrary_image_file(Resources.YW5003_5)
        small_plot_island = ami_graph.get_or_create_ami_islands(mindim=50, maxmindim=300)[0]
        ami_edges = small_plot_island.get_or_create_ami_edges()
```
we need to set tolerances as the pixels can be +- 1 or even more from the mean. First we find the horiz/vert edges:
```
        vertical_edges = AmiEdge.get_vertical_edges(ami_edges, tolerance=2)
        assert len(vertical_edges) == 29
        horizontal_edges = AmiEdge.get_horizontal_edges(ami_edges, tolerance=2)
        assert len(horizontal_edges) == 5
```
then we extract the `AmiLine`s.  An edge will only have a line if all the pixels (points) in it are within `tolerance` 
```
        vert_ami_lines = self.get_single_lines(vertical_edges)
        assert len(vert_ami_lines) == 29
        horiz_ami_lines = self.get_single_lines(horizontal_edges)
        assert len(horiz_ami_lines) == 5
```
Now we create Counters(dictionaries) based on the perpendicular coordinate of the line (i.e. x for vertical, y for horizontal). Thus 12 vertical lines have an `x-coord = 66`, 4 horizontal have `y=759`. We'll need to aggregate close values (e.g. 758 and 759)
```
        vert_dict = AmiLine.get_horiz_vert_counter(vert_ami_lines, xy_index=0)
        assert vert_dict == Counter({66: 12, 295: 11, 65: 1, 107: 1, 149: 1, 191: 1, 232: 1, 274: 1}),\
            f"found {vert_dict}"
        horiz_dict = AmiLine.get_horiz_vert_counter(horiz_ami_lines, xy_index=1)
        assert horiz_dict == Counter({759: 4, 758: 1}), f"found {horiz_dict}"

```

**Comment by anuvc** [comment]
### AmiOCR extracts words and removes special characters
I've extracted the text and boxes from Biosynth2 pathway diagram. There are some minor mistakes that needs fixing. All of the text has been recognised and the special characters have been removed. 

![biosynth_path_2](https://user-images.githubusercontent.com/45652303/150679108-3d281120-0496-4215-99fd-4fc107f809af.jpg)

```
text: Glycolysis bbox: [[405, 638], [1, 57]]
text: Terpene bbox: [[182, 349], [57, 99]]
text: Biosynthetic bbox: [[140, 390], [111, 145]]  
text: Bethea bbox: [[178, 329], [122, 200]]        
text: Acetyl-Co bbox: [[606, 798], [149, 187]]     
text: Biosynthetic bbox: [[1004, 1253], [127, 177]]
text: carboxylase bbox: [[606, 805], [196, 235]]   
text: Pathway bbox: [[1042, 1217], [186, 229]]     
text: Acetyl-Co bbox: [[413, 613], [230, 267]]
text: Malonyl-Co bbox: [[793, 1022], [232, 270]]
text: Acetoacetyl-Co bbox: [[198, 479], [279, 344]]
text: -ketoacyl-A bbox: [[926, 1196], [301, 338]]
text: synthase bbox: [[1205, 1350], [301, 338]]
text: synthase bbox: [[268, 414], [330, 368]]
text: Acetoacetyl-Co bbox: [[373, 658], [380, 427]]
text: Acetoacetyl-Co bbox: [[772, 1058], [384, 422]]
text: 3-Hydroxy-3-met bbox: [[133, 487], [429, 476]]
text: f-ketoacy bbox: [[931, 1107], [452, 490]]
text: reductase bbox: [[1117, 1278], [452, 481]]
text: glutaryl-CoA-syntha bbox: [[118, 500], [482, 520]]
text: 3-Hydroxy-3-met bbox: [[356, 708], [540, 578]]
text:  B-hydroxy-butyryl bbox: [[771, 1166], [537, 575]]
text: glutaryl-Co bbox: [[424, 640], [588, 626]]
text: Mevalonate bbox: [[293, 491], [632, 679]]
text: DHA bbox: [[1076, 1181], [75, 109]]
text: B-hydroxyacy bbox: [[933, 1176], [608, 646]]
text: dehydratase bbox: [[1186, 1387], [608, 646]]
text: synthase bbox: [[319, 465], [684, 722]]
text: Crotonyl-AC bbox: [[801, 1038], [699, 737]]
text: Mevalonate bbox: [[426, 618], [735, 764]]
text: Mevalonate bbox: [[295, 494], [773, 820]]
text: Enoyl bbox: [[939, 1043], [775, 813]]
text: reductase bbox: [[1052, 1213], [775, 804]]
text: kinase bbox: [[341, 450], [825, 854]]
text: Mevalonate-5-phospha bbox: [[332, 738], [877, 924]]
text: Butyryl bbox: [[816, 940], [849, 927]]
text: ac bbox: [[954, 1033], [853, 883]]
text: Mevalonate- bbox: [[223, 469], [925, 970]]
text: phosphate-kinas bbox: [[193, 494], [972, 1019]]
text: Mevalonate-5-pyrophospha bbox: [[298, 780], [1025, 1063]]
text: Docosahexanoic bbox: [[823, 1092], [1029, 1059]]
text: acid bbox: [[1106, 1173], [1029, 1059]]
text: Isopentenyl bbox: [[197, 407], [1077, 1115]]
text: pyrophosphate bbox: [[96, 349], [1125, 1162]]
text: synthase bbox: [[359, 505], [1125, 1162]]
text: Isopentenyl bbox: [[317, 519], [1174, 1220]]
text: See bbox: [[521, 761], [1178, 1257]]
text: Farnesyl bbox: [[221, 384], [1221, 1259]]
text: pyrophosphate bbox: [[97, 349], [1265, 1310]]
text: synthase bbox: [[359, 526], [1247, 1306]]
text: Squalene bbox: [[703, 860], [1265, 1303]]
text: synthase bbox: [[709, 854], [1312, 1350]]
text: B-caroten bbox: [[6, 177], [1347, 1385]]
text: Farnesyl-pyrophosphat bbox: [[342, 736], [1350, 1388]]
text: Squalene bbox: [[851, 1174], [1353, 1391]]
text: Cholesterol bbox: [[1231, 1419], [1357, 1388]]
```

**Reply by petermr** [reply]
Thanks,
Is the image linked from the text? I can't see it. [Uploading biosynth_path_2.jpg…]

**Reply by anuvc** [reply]
Yes, the image didn't properly upload the last time. I've fixed it now.

**Comment by anuvc** [comment]
### Tesseract OCR options
There seems to be information about the baseline of words in a line in <ocr_line> property tesseract.
The line has baseline, we can calculate the base line for each of the words, using the x values and slope. 

There also are ascender and descender values, and font size.
We can find information regarding the properties here: http://kba.cloud/hocr-spec/1.2/#baseline

![image](https://user-images.githubusercontent.com/45652303/150722979-6277ab4f-3fbf-4918-8b35-0a75fb96908d.png)

**Comment by roopavsm** [comment]
The goals that we need to set for commandline are:

1. parameters for processing images:

- octree

- grayscale

-  erode-dilate

- thinning/skeleton

2. lines and edges

- nx_graph

- segmentation

- polylines

- polygons

3. text

- tesseract

- phrase assembly

- error correction

4. higher primitives

- plotbox

   a. spines
   b. limits

- scales

   a. ticks
   b. values
   c. labels
   d. units
   e. pixel2 user coords

- arrows

- text-boxes

We need Github pages, examples, tests,  etc for this

- commandline for all of this.

**Comment by petermr** [comment]
Thanks,
@anuv will manage and assemble this
I will create commandline options in `nodes_and_pixels` for:
* initial processing `--image`
* lines_and_edges `--line` and `--edge`
* arrows `--arrow`
* (plotbox and scales) - not really relevant for reactions


I will stay clear of
* `text` ***@***.***)
* errors

My intention is that my commands can be developed in `nodes_and_pixels `
and then integrated with minimal fuss.

Rather than have huge numbers of commands I expect to create a dict-like
approach:

`imageananalysis --edge maxdim=100 delta_seg=2  ... ` It means that my
commands take a list of name-value pairs which is flexible, can be
configured from files, but I have to write a simple parse. I did this for
py4ami.







-- 
Peter Murray-Rust
Founder ContentMine.org
and
Reader Emeritus in Molecular Informatics
Dept. Of Chemistry, University of Cambridge, CB2 1EW, UK

**Reply by anuvc** [reply]
I think we should maintain the command line file in the main branch together because It'll be only one file for accessing the command line

I made a file `pyimage/pyamiimage.py` which does basic CLI for text (for now) and I've committed it to main branch.

**Comment by petermr** [comment]
thanks,
Which file do you suggest? it should probably be the name of the program
(e.g. imageanalysis.py)

On Fri, Feb 18, 2022 at 5:30 AM Anubhab Chakraborty <
***@***.***> wrote:

> I think we should maintain the command line file in the main branch
> together because It'll be only one file for accessing the command line
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/petermr/petermr/discussions/8#discussioncomment-2202040>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFTCS7LOFIP2EA6ER3Q4FDU3XKOVANCNFSM5LRVZBWQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Peter Murray-Rust
Founder ContentMine.org
and
Reader Emeritus in Molecular Informatics
Dept. Of Chemistry, University of Cambridge, CB2 1EW, UK

**Reply by anuvc** [reply]
We've set the name of the program as `pyamiimage`

**Comment by anuvc** [comment]
Yay! We have published a version of `pyamiimage` that works! Current functionality is limited to OCR. We will be adding more functionality in the coming weeks. Right now you can install using:

```pip install --upgrade pyamiimage```

Run it using:
```pyamiimage --help```

Changes in the update `0.0.12`:
- `pyamiimage` works in the command line, and can also be imported as a module
- Imports have been fixed since version `0.0.9`
- specific package versions requirements has been removed to support latest version of packages
- File structure has been improved, so its much easier to understand and contribute

**Comment by petermr** [comment]
Thanks very much!!

On Tue, Mar 15, 2022 at 5:54 AM Anubhab Chakraborty <
***@***.***> wrote:

> Yay! We have published a version of pyamiimage that works! Current
> functionality is limited to OCR. We will be adding more functionality in
> the coming weeks. Right now you can install using:
>
> pip install --upgrade pyamiimage
>
> Run it using:
> pyamiimage --help
>
> Changes in the update 0.0.12:
>
>    - pyamiimage works in the command line, and can also be imported as a
>    module
>    - Imports have been fixed since version 0.0.9
>    - specific package versions requirements has been removed to support
>    latest version of packages
>    - File structure has been improved, so its much easier to understand
>    and contribute
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/petermr/petermr/discussions/8#discussioncomment-2362554>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFTCS7UPH5ROVZGFPK6LJLVAAQZLANCNFSM5LRVZBWQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Peter Murray-Rust
Founder ContentMine.org
and
Reader Emeritus in Molecular Informatics
Dept. Of Chemistry, University of Cambridge, CB2 1EW, UK

**Comment by petermr** [comment]
# extracting lines from plots

See test_plots.py:

```
    def test_example_042a(self):
        """creates AmiPlot and reads image and creates scaled data
        """
        # input file
        path042a = Path(Resources.TEST_RESOURCE_DIR, "042A", "raw.png")
        ami_plot = AmiPlot(image_file=path042a)
        ami_plot.create_calibrated_plot_box()
        assert ami_plot.bottom_scale.user_to_plot_scale == 19.575
        edge_points_xy = ami_plot.get_points_xy_for_single_edge()
        assert 1150 < len(edge_points_xy) < 1160, f" got {len(edge_points_xy)} instead of 1158"
        points_user_xy = ami_plot.scale_plot_points_to_user(edge_points_xy)
        path = Path(Resources.TEMP_DIR, "042a_user_coords.csv")
        AmiUtil.write_xy_to_csv(points_user_xy, path, ['X', 'Y'])
```
This creates an X-Y plot which compares precisely with the original

<img width="1385" alt="Screenshot 2022-04-03 at 19 46 46" src="https://user-images.githubusercontent.com/733515/161443757-3d494b55-7cda-42e9-bb22-a0b097fbd667.png">

**Reply by anuvc** [reply]
Looks great! I tested it out on my system. It works!
